library/src/G4CMPPhononBoundaryProcess.cc

```diff
@@ -192,128 +227,249 @@ DoReflection(const G4Track& aTrack, const G4Step& aStep,
+    // (negative implied in kPerpMag for inward pointing)
+    kPerpV = kPerpMag * newNorm;
+
+    // Calculate new reflectedKDir (kTan + kPerpV) and Vg
+    reflectedKDir = kTan + kPerpV;
+    G4ThreeVector vDir = theLattice->MapKtoVDir(mode, reflectedKDir);
+
+    if (verboseLevel>3) {
+      G4cout << " GetSpecularVector:insideLoop -> "
+	     << " attempts = " << nAttempts
+	     << ", oldstepLocalPos = " << oldstepLocalPos/mm << " mm"
+	     << ", stepLocalPos = " << stepLocalPos/mm << " mm"
+	     << ", oldNorm = " << oldNorm
+	     << ", newNorm = " << newNorm
+	     << ", kPerpMag = " << kPerpMag
+	     << ", kPerpV (kPerpMag * newNorm) = " << kPerpV
+	     << ", kTan = " << kTan
+	     << ", reflectedKDir (kTan + kPerpV) = " << reflectedKDir
+	     << ", Phonon mode = " << G4PhononPolarization::Label(mode)
+	     << ", New group velocity: " << vDir << G4endl;
     }
-    
-    (reflectedKDir -= kstep*surfNorm).setMag(1.);
-    olddir = newdir;
-    nstep++;
-  } 
-  
-  if (nstep>0 && verboseLevel) {
-    G4cout << " adjusted specular reflection with " << nstep << " steps"
-	   << " (" << nflip << " flips) kPerp " << kPerp << G4endl;
   }
 
-  return reflectedKDir;
-}
+  // Restore global coordinates to new vectors
+  RotateToGlobalDirection(reflectedKDir);
+  RotateToGlobalDirection(newNorm);
+  RotateToGlobalPosition(stepLocalPos);
+
+  if (!G4CMP::PhononVelocityIsInward(theLattice, mode, reflectedKDir, newNorm,
+                                     stepLocalPos)) {
+    if (verboseLevel) {
+      G4cerr << GetProcessName() << "::GetSpecularVector"
+	     << ": Phonon displacement failed after " << nAttempts - 1
+	     << " attempts." << G4endl;
+      if (verboseLevel>1) {
+	G4cout << "Doing diffuse reflection at surface point " 
+	       << surfacePoint/mm << " mm" << G4endl;
+      }
+    }
 
+    // Get reflectedKDir from initial point and restore original values
+    stepLocalPos = surfacePoint;
+    newNorm = surfNorm;
+    reflectedKDir = G4CMP::GetLambertianVector(theLattice, surfNorm, mode,
+                                               surfacePoint);
+  }
 
-// Generate diffuse reflection according to 1/cos distribution
+  if (verboseLevel>3) {
+    G4cout << GetProcessName() << "::GetSpecularVector"
+	   << ": nAttempts = " << nAttempts
+	   << ", waveVector = " << waveVector
+	   << ", reflectedKDir = " << reflectedKDir
+	   << ", initialGlobalPostion = " << surfacePoint/mm << " mm"
+	   << ", finalGlobalPosition = " << stepLocalPos/mm << " mm"
+	   << G4endl;
+  }
 
-G4ThreeVector G4CMPPhononBoundaryProcess::
-GetLambertianVector(const G4ThreeVector& surfNorm, G4int mode) const {
-  G4ThreeVector reflectedKDir;
-  const G4int maxTries = 1000;
-  G4int nTries = 0;
-  do {
-    reflectedKDir = G4CMP::LambertReflection(surfNorm);
-  } while (nTries++ < maxTries &&
-	   !G4CMP::PhononVelocityIsInward(theLattice, mode,
-					  reflectedKDir, surfNorm));
+  delete solidUtils;
```

Comment:
If you make `solidUtils` a simple local variable, you don't need to this delete.  Also, if the function breaks/returns in the middle for some reason, it'll also be cleaned up automatically.